core/shard/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+extern crate chain;
+extern crate parking_lot;
+extern crate primitives;
+extern crate rand;
+#[macro_use]
+extern crate serde_derive;
+extern crate storage;
+
+use chain::{Block, Header};
+use primitives::hash::{CryptoHash, hash_struct};
+use primitives::types::{AuthorityMask, MerkleHash, MultiSignature, PartialSignature, ReceiptTransaction, SignedTransaction};
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct ShardBlockHeaderBody {
+    pub parent_hash: CryptoHash,
+    pub index: u64,
+    pub merkle_root_state: MerkleHash,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct ShardBlockHeader {
+    pub body: ShardBlockHeaderBody,
+    pub hash: CryptoHash,
+    pub authority_mask: AuthorityMask,
+    pub signature: MultiSignature,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct ShardBlockBody {
+    pub header: ShardBlockHeaderBody,
```

Comment:
I also suggest to not put header inside the body. Because if we anywhere have a vector (or other container) of bodies and then we decide to have a vector of headers then we can only either:
1) have vector of references to the headers
2) or vector of cloned headers

Reply:
we would not have a vector of headers separately. It's either going to be inside SignedShardBlock or SignedShardBlockHeader, so copy will be inevitable if you have vector of blocks and headers.